clippy_lints/src/methods/map_clone.rs

```diff
@@ -119,6 +122,9 @@ fn handle_path(
             && let args = args.as_slice()
             && let Some(ty) = args.iter().find_map(|generic_arg| generic_arg.as_type())
             && ty.is_ref()
+            && let ty::Ref(_, ty, Mutability::Not) = ty.kind()
+            && let ty::FnDef(_, lst) = cx.typeck_results().expr_ty(arg).kind()
```

Comment:
`cx.typeck_results()` returns the type check results for the current body. This is a paper cut, with this lint, since the code of the Closure is considered a different body. There are two possible fixes for this:
1. We could get the type check results for the body of the closure and do the comparison you're doing here. I believe that the IDs of generic types could mismatch if we take this approach. In the worst case, this would result in false negatives(NFs) which are better than the false positive(FP) you're fixing here.
2. The second option, is to get the type of the `map` expression (argument `e` of this function), which is inside the same body as the receiver. Then we can extract the generic type of that iterator, like it's done for the receiver. This should be sufficient, with the check you added above.

    ```rs
    let x: Option<&u8> = None;
    //            vvvvvvvvvvvvvvvvvvvvvvvvvv This is a new body
    //                              vvvvvvv `arg`
    let y = x.map(|x| String::clone(loop {}));
    //      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ e ~= Iterator<String>
    //      ^ recv ~= Iterator<&u32>
    ```

I favor option 2 as it seems simpler and more resilient.

Does this explanation give you enough information, to try either solution?

Reply:
Thank you for your review! The option 2 makes sense for me.